metricbeat/module/system/raid/blockinfo/parser.go

```diff
@@ -0,0 +1,193 @@
+	size, err := parseIntVal(filepath.Join(path, "size"))
+	if err != nil {
+		return Disk{}, errors.Wrap(err, "error getting disk size")
+	}
+
+	state, err := ioutil.ReadFile(filepath.Join(path, "state"))
+	if err != nil {
+		return Disk{}, errors.Wrap(err, "error getting disk state")
+	}
+
+	return Disk{Size: size, State: strings.TrimSpace(string(state))}, nil
+
+}
+
+//get the current sync status as it exists under md/sync_completed
+//if there's no sync operation in progress, the file will just have 'none'
+//in which case, default to to the overall size
+func getSyncStatus(path string, size int64) (SyncStatus, error) {
+
+	re := regexp.MustCompile(`^(\d*) \/ (\d*)`)
+
+	raw, err := ioutil.ReadFile(filepath.Join(path, "md", "sync_completed"))
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "could not open sync_completed")
+	}
+	if string(raw) == "none\n" {
+		return SyncStatus{Complete: size, Total: size}, nil
+	}
+
+	matches := re.FindStringSubmatch(string(raw))
+
+	if len(matches) < 3 {
+		return SyncStatus{}, fmt.Errorf("Could not get data from sync_completed")
+	}
+
+	current, err := strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "Could not parse data sync_completed")
+	}
+
+	total, err := strconv.ParseInt(matches[2], 10, 64)
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "Could not parse data sync_completed")
+	}
+
+	return SyncStatus{Complete: current, Total: total}, nil
+
+}
+
+//Create a new disk object, parsing any needed fields
+func newMD(path string) (MDDevice, error) {
+	var dev MDDevice
+
+	dev.Name = filepath.Base(path)
+	size, err := parseIntVal(filepath.Join(path, "size"))
+	if err != nil {
+		return dev, errors.Wrap(err, "could not get device size")
+	}
+	dev.Size = size
+
+	//This is the count of 'active' disks
+	active, err := parseIntVal(filepath.Join(path, "md", "raid_disks"))
+	if err != nil {
+		return dev, errors.Wrap(err, "could not get raid_disks")
+	}
+	dev.ActiveDisks = active
+
+	//RAID array state
+	state, err := ioutil.ReadFile(filepath.Join(path, "md", "array_state"))
+	if err != nil {
+		return dev, errors.Wrap(err, "could not open array_state")
+	}
+	dev.ArrayState = strings.TrimSpace(string(state))
+
+	//get total disks
+	disks, err := getDisks(path)
+	if err != nil {
+		return dev, errors.Wrap(err, "could not get disk data")
+	}
+	dev.Devices = disks
+
+	//sync action and sync status will only exist for redundant raid levels
+	redundant, err := isRedundant(path)
+	if err != nil {
+		return dev, errors.Wrap(err, "error getting raid level")
+	}
+	if redundant {
+
+		//Get the sync action
+		//Will be idle if nothing is going on
+		syncAction, err := ioutil.ReadFile(filepath.Join(path, "md", "sync_action"))
+		if err != nil {
+			return dev, errors.Wrap(err, "could not open sync_action")
+		}
+		dev.SyncAction = strings.TrimSpace(string(syncAction))
+
+		//sync status
+		syncComp, err := getSyncStatus(path, dev.Size)
```

Comment:
What is "comp" here? could we call this variable `syncStatus`?

Reply:
"sync completed". Should probably make it more clear.